Opam init and next steps failing with ARCH NOT VALID VAR

Howdy. Opam init fails with

> [ERROR] arch is not a valid variable.

Also all the next ones as here:
`
butt:~ lykhovyd $ opam update
default    Downloading https://opam.ocaml.org/urls.txt
Updating ~/.opam/repo/compiler-index ...
Updating ~/.opam/compilers/ ...
Updating ~/.opam/repo/package-index ...
Updating ~/.opam/packages/ ...
Updating the cache of metadata (~/.opam/state.cache) ...
[ERROR] arch is not a valid variable.
`
My system:
`
uname -a
Linux butt 3.16-3-amd64 #1 SMP Debian 3.16.5-1 (2014-10-10) x86_64 GNU/Linux
`

--debug output

> building a lot of staff here
> 
> make[2]: Leaving directory '/home/lykhovyd/.opam/4.02.1/build/ocaml/tools'
> make[1]: Leaving directory '/home/lykhovyd/.opam/4.02.1/build/ocaml'
> 03:43.286  034098  SYSTEM                  [log-34098-1b434b](in 0.000s) make install
> 03:43.286  034098  FILENAME                rmdir /home/lykhovyd/.opam/4.02.1/build/ocaml
> 03:43.286  034098  SYSTEM                  rmdir /home/lykhovyd/.opam/4.02.1/build/ocaml
> 03:43.401  034098  STATE                   add_switch switch=4.02.1 compiler=4.02.1
> 03:43.403  034098  FILE(config)            Wrote ~/.opam/config in 0.000s
> 03:43.404  034098  SYSTEM                  mkdir /home/lykhovyd/.opam/opam-init
> 03:43.404  034098  CLIENT                  installing compiler packages
> 03:43.404  034098  STATE                   LOAD-STATE(init-3)
> 03:43.443  034098  FILE(installed)         Cannot find /home/lykhovyd/.opam/4.02.1/installed
> 03:43.443  034098  FILE(pinned)            Cannot find /home/lykhovyd/.opam/4.02.1/pinned
> 03:43.444  034098  FILE(installed.roots)   Cannot find /home/lykhovyd/.opam/4.02.1/installed.roots
> 03:43.875  034098  FILE(reinstall)         Cannot find /home/lykhovyd/.opam/4.02.1/reinstall
> 03:43.875  034098  STATE                   ROOT      : /home/lykhovyd/.opam
> 03:43.875  034098  STATE                   SWITCH    : 4.02.1
> 03:43.875  034098  STATE                   COMPILER  : 4.02.1
> 03:43.875  034098  STATE                   COMPILERS : { 3.11.2, 3.12.1, 4.00.0, 4.00.0+debug-runtime, 4.00.0+fp, 4.00.1, 4.00.1+BER, 4.00.1+alloc-profiling, 4.00.1+annot, 4.00.1+french, 4.00.1+mirage-unix, 4.00.1+mirage-xen, 4.00.1+open-types, 4.00.1+raspberrypi, 4.00.1+short-types, 4.01.0, 4.01.0+32bit, 4.01.0+BER, 4.01.0+PIC, 4.01.0+armv6-freebsd, 4.01.0+bin-ocp, 4.01.0+fp, 4.01.0+jocaml, 4.01.0+lsb, 4.01.0+musl, 4.01.0+musl+static, 4.01.0+open-types, 4.01.0+profile, 4.02.0, 4.02.0+PIC, 4.02.0+improved-errors, 4.02.0+modular-implicits, 4.02.0+rc1, 4.02.0+trunk, 4.02.1, 4.02.1+32bit, 4.02.1+PIC, 4.03.0+fp, 4.03.0+pr100, 4.03.0+pr101, 4.03.0+pr102, 4.03.0+pr103, 4.03.0+pr107, 4.03.0+pr108, 4.03.0+pr109, 4.03.0+pr43, 4.03.0+pr46, 4.03.0+pr51, 4.03.0+pr57, 4.03.0+pr63, 4.03.0+pr65, 4.03.0+pr68, 4.03.0+pr69, 4.03.0+pr71, 4.03.0+pr73, 4.03.0+pr76, 4.03.0+pr78, 4.03.0+pr79, 4.03.0+pr80, 4.03.0+pr84, 4.03.0+pr88, 4.03.0+pr89, 4.03.0+pr92, 4.03.0+pr93, 4.03.0+pr94, 4.03.0+pr95, 4.03.0+pr96, 4.03.0+pr98, 4.03.0+trunk, system }
> 03:43.875  034098  STATE                   REPOS     : { default }
> 03:43.875  034098  STATE                   PACKAGES  : 2825 packages
> 03:43.875  034098  STATE                   INSTALLED : {}
> 03:43.875  034098  STATE                   ROOTS     : {}
> 03:43.875  034098  STATE                   REINSTALL : {}
> Creating a cache of metadata in ~/.opam/state.cache ...
> 03:43.887  034098  STATE                   ~/.opam/state.cache written in 0.012s
> 03:43.887  034098  STATE                   State init-3 loaded in 0.480s
> 03:43.891  034098  FILE(.config)           Cannot find /home/lykhovyd/.opam/4.02.1/lib/camlp4/opam.config
> 03:43.891  034098  FILE(.config)           Cannot find /home/lykhovyd/.opam/4.02.1/lib/camlp4/opam.config
> 03:43.891  034098  FILE(.config)           Cannot find /home/lykhovyd/.opam/4.02.1/lib/camlp4/opam.config
> 03:43.892  034098  FILE(.config)           Cannot find /home/lykhovyd/.opam/4.02.1/lib/camlp4/opam.config
> 03:43.892  034098  FILE(.config)           Cannot find /home/lykhovyd/.opam/4.02.1/lib/camlp4/opam.config
> 03:43.892  034098  FILE(.config)           Cannot find /home/lykhovyd/.opam/4.02.1/lib/camlp4/opam.config
> 03:43.892  034098  FILE(.config)           Cannot find /home/lykhovyd/.opam/4.02.1/lib/camlp4/opam.config
> 03:43.892  034098  FILE(.config)           Cannot find /home/lykhovyd/.opam/4.02.1/lib/camlp4/opam.config
> 03:43.892  034098  FILE(.config)           Cannot find /home/lykhovyd/.opam/4.02.1/lib/camlp4/opam.config
> 03:43.893  034098  FILE(.config)           Cannot find /home/lykhovyd/.opam/4.02.1/lib/camlp4/opam.config
> 03:43.893  034098  FILE(.config)           Cannot find /home/lykhovyd/.opam/4.02.1/lib/camlp4/opam.config
> 03:43.893  034098  FILE(.config)           Cannot find /home/lykhovyd/.opam/4.02.1/lib/camlp4/opam.config
> 03:43.893  034098  FILE(.config)           Cannot find /home/lykhovyd/.opam/4.02.1/lib/camlp4/opam.config
> 03:43.894  034098  FILE(.config)           Cannot find /home/lykhovyd/.opam/4.02.1/lib/camltc/opam.config
> 03:43.894  034098  FILE(.config)           Cannot find /home/lykhovyd/.opam/4.02.1/lib/camltc/opam.config
> 03:43.894  034098  FILE(.config)           Cannot find /home/lykhovyd/.opam/4.02.1/lib/camltc/opam.config
> 03:43.894  034098  FILE(.config)           Cannot find /home/lykhovyd/.opam/4.02.1/lib/camltc/opam.config
> 03:43.894  034098  FILE(.config)           Cannot find /home/lykhovyd/.opam/4.02.1/lib/camltc/opam.config
> 03:43.894  034098  FILE(.config)           Cannot find /home/lykhovyd/.opam/4.02.1/lib/camltc/opam.config
> 03:43.897  034098  FILE(.config)           Cannot find /home/lykhovyd/.opam/4.02.1/lib/core_kernel/opam.config
> [ERROR] arch is not a valid variable.
> 'opam init --comp=4.02.1 --debug' failed.

--- a/admin-scripts/to_1_1.ml
+++ b/admin-scripts/to_1_1.ml
@@ -8,18 +8,56 @@
    * remove dependency flags ('build', 'test', 'doc')
    * set file version
    * replace inequality constraints with '> & <'
+   * remove new global variables from filters in commands, messages, 'patches',
+     'available'
 *)
+
+open OpamTypes
+open OpamMisc.Option.Op
+;;
 
 OpamGlobals.all_parens := true;;
 
 let rewrite_constraint ~conj = (* Rewrites '!=' *)
   OpamFormula.map OpamFormula.(function
       | (`Neq,v) ->
-        prerr_endline "XX";
         if conj then And (Atom (`Lt,v), Atom (`Gt,v))
         else Or (Atom (`Lt,v), Atom (`Gt,v))
       | atom -> Atom atom)
 ;;
+
+let vars_new_1_2 = [ "compiler"; "ocaml-native"; "ocaml-native-tools";
+                     "ocaml-native-dynlink"; "arch" ]
+
+let rec filter_vars = function
+  | FIdent i when List.mem i vars_new_1_2 -> None
+  | FBool _ | FString _ | FIdent _ as f -> Some f
+  | FOp (f1,op,f2) ->
+     (match filter_vars f1, filter_vars f2 with
+      | Some f1, Some f2 -> Some (FOp (f1, op, f2))
+      | _ -> None)
+  | FAnd (f1,f2) ->
+     (match filter_vars f1, filter_vars f2 with
+      | Some f1, Some f2 -> Some (FAnd (f1, f2))
+      | opt, None | None, opt -> opt)
+  | FOr (f1,f2) ->
+     (match filter_vars f1, filter_vars f2 with
+      | Some f1, Some f2 -> Some (FOr (f1, f2))
+      | opt, None | None, opt -> opt)
+  | FNot f ->
+     (match filter_vars f with
+      | Some f -> Some (FNot f)
+      | None -> None)
+
+let filter_vars_optlist ol =
+  List.map
+    (fun (x, filter) -> x, filter >>= filter_vars)
+    ol
+
+let filter_vars_commands ol =
+  List.map
+    (fun (args, filter) -> filter_vars_optlist args, filter >>= filter_vars)
+    ol
 
 let to_1_1 _ opam =
   let module OF = OpamFile.OPAM in
@@ -28,7 +66,9 @@ let to_1_1 _ opam =
     < 0
   then opam
   else
-  let opam = OF.with_build opam (OF.build opam @ OF.install opam) in
+  let opam =
+    OF.with_build opam
+      (filter_vars_commands (OF.build opam @ OF.install opam)) in
   let opam = OF.with_install opam [] in
   let opam = OF.with_flags opam [] in
   let opam = OF.with_dev_repo opam None in
@@ -45,6 +85,13 @@ let to_1_1 _ opam =
            OpamFormula.Atom (n, rewrite_constraint ~conj:true cstr))
           (OF.conflicts opam))
   in
+  let opam = OF.with_available opam (filter_vars (OF.available opam) +! FBool true) in
+  let opam = OF.with_patches opam (filter_vars_optlist (OF.patches opam)) in
+  let opam = OF.with_libraries opam [] in
+  let opam = OF.with_syntax opam [] in
+  let opam = OF.with_messages opam (filter_vars_optlist (OF.messages opam)) in
+  let opam =
+    OF.with_post_messages opam (filter_vars_optlist (OF.post_messages opam)) in
   opam
 ;;
 

--- a/src/core/opamFile.ml
+++ b/src/core/opamFile.ml
@@ -1137,10 +1137,13 @@ module X = struct
     let with_substs t substs = { t with substs }
     let with_ocaml_version t ocaml_version = { t with ocaml_version }
     let with_os t os = { t with os }
+    let with_available t available = { t with available }
     let with_maintainer t maintainer = { t with maintainer }
     let with_patches t patches = { t with patches }
     let with_bug_reports t bug_reports = { t with bug_reports }
     let with_depexts t depexts = { t with depexts }
+    let with_messages t messages = { t with messages }
+    let with_post_messages t post_messages = { t with post_messages }
     let with_flags t flags = { t with flags }
     let with_dev_repo t dev_repo = {t with dev_repo }
 

--- a/src/core/opamFile.mli
+++ b/src/core/opamFile.mli
@@ -161,6 +161,10 @@ module OPAM: sig
   (** External dependencies *)
   val depexts: t -> tags option
 
+  val with_messages: t -> (string * filter option) list -> t
+
+  val with_post_messages: t -> (string * filter option) list -> t
+
   (** Package conflicts *)
   val conflicts: t -> formula
 
@@ -249,6 +253,8 @@ module OPAM: sig
   val with_ocaml_version: t -> compiler_constraint option -> t
 
   val with_os: t -> (bool * string) generic_formula -> t
+
+  val with_available : t -> filter -> t
 
   (** Construct as [maintainer] *)
   val with_maintainer: t -> string list -> t

